File: Semestrul 5/PPL/Laborator1/src/main/java/ro/ppd2024/distributor/LinearDistributor.java
package ro.ppd2024.distributor;


import ro.ppd2024.thread.linear.DeltaLinearConvolutionThread;
import ro.ppd2024.thread.linear.HorizontalLinearConvolutionThread;
import ro.ppd2024.thread.linear.VerticalLinearConvolutionThread;
import ro.ppd2024.util.Technique;

public class LinearDistributor {
    public static void distribute(int[][] F, int n, int m, int[][] V, int[][] C, int k,
                                  int nrThreads, Technique technique) {
        int size = switch (technique) {
            case HORIZONTAL_LINEAR -> n / nrThreads;
            case VERTICAL_LINEAR -> m / nrThreads;
            case DELTA_LINEAR -> n * m / nrThreads;
            default -> throw new IllegalArgumentException("Invalid technique");
        };

        int remainder = n % nrThreads;
        Thread[] threads = new Thread[nrThreads];
        int start = 0, end, i = 0;

        while (start < n) {
            end = start + size;
            if (remainder > 0) {
                end++;
                remainder--;
            }

            switch (technique) {
                case HORIZONTAL_LINEAR:
                    threads[i] = new HorizontalLinearConvolutionThread(F, V, n, m, C, k, start, end);
                    break;
                case VERTICAL_LINEAR:
                    threads[i] = new VerticalLinearConvolutionThread(F, V, n, m, C, k, start, end);
                    break;
                case DELTA_LINEAR:
                    threads[i] = new DeltaLinearConvolutionThread(F, V, n, m, C, k, start, end);
                    break;
            }
            threads[start].start();
            start = end;
            i++;
        }

        for (i = 0; i < nrThreads; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

File: Semestrul 5/PPD/Laborator1/ConvolutionCpp/distributor/LinearDistributor.cpp
#include "LinearDistributor.h"
#include "../thread/LinearConvolutionThread.h"
#include <stdexcept>
#include <thread>

template <typename TMatrix, typename TFilter>
void LinearDistributor<TMatrix, TFilter>::distribute(TMatrix F, int n, int m, TMatrix V, TFilter C, int k, int nrThreads, Technique technique) {
    int limit;
    switch (technique) {
        case Technique::HORIZONTAL_LINEAR:
            limit = n;
            break;
        case Technique::VERTICAL_LINEAR:
            limit = m;
            break;
        case Technique::DELTA_LINEAR:
            limit = n * m;
            break;
        default:
            throw std::invalid_argument("Invalid technique");
    }

    nrThreads = std::min(nrThreads, limit);
    int size = limit / nrThreads;
    int remainder = limit % nrThreads;
    int start = 0, end, i = 0;
    std::thread* threads = new std::thread[nrThreads];

    while (start < limit) {
        end = start + size;
        if (remainder > 0) {
            end++;
            remainder--;
        }

        switch(technique) {
            case Technique::HORIZONTAL_LINEAR:
                threads[i] = std::thread(horizontalLinearConvolutionThread, F, V, n, m, C, k, start, end);
                break;
            case Technique::VERTICAL_LINEAR:
                threads[i] = std::thread(verticalLinearConvolutionThread, F, V, n, m, C, k, start, end);;
                break;
            case Technique::DELTA_LINEAR:
                threads[i] = std::thread(deltaLinearConvolutionThread, F, V, n, m, C, k, start, end);;
                break;
        }
        start = end;
        i++;
    }

    for (i = 0; i < nrThreads; i++) {
        threads[i].join();
    }
    delete[] threads;
}